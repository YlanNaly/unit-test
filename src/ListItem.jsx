export function ListItem(props) {
    const {checkable, onCheck, id, item} = props;

    return <li className="list-item" data-testid={`test-${id}-container`}>
        {checkable && (
            <input  type="checkbox" id={id} data-testid={`test-${id}`}
                   onChange={checkable ? 
                    (event, item, i) => onCheck(event, item, i) : 
                    () => {}
                }/>
        )}
        <label data-testid="label-item" htmlFor={id}>{item}</label>
    </li>;
}