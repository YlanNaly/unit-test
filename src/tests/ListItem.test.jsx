import {fireEvent ,screen, render} from "@testing-library/react";

import { ListItem } from "../ListItem";

const mockOnCheck = jest.fn();

describe('ListItem', () => {
    it('display value correctly', () => {
        const { getByText } = render(
            <ListItem
                id='list-item-test'
                checkable={true}
                onCheck={mockOnCheck}
                item='Lorem ipsum dolor sit amet consectetur'
            />
        );

        const value = getByText('Lorem ipsum dolor sit amet consectetur');
        expect(value).toBeInTheDocument();
    });
    
    it('checkbox is shown', () => {
        const { getByTestId } = render(
            <ListItem
                id='list-item-1'
                checkable={true}
                onCheck={mockOnCheck}
                item='Lorem ipsum dolor sit amet consectetur'
            />
        );

        const value = getByTestId('test-list-item-1');
        expect(value).toBeInTheDocument();
    });

    it('checkbox is hidden', () => {
        const { getByTestId, debug } = render(
            <ListItem
                id='list-item-1'
                checkable={false}
                onCheck={mockOnCheck}
                item='Lorem ipsum dolor sit amet consectetur'
            />
        );
        const node = getByTestId('test-list-item-1-container');
        expect(node.children).toHaveLength(1);
    });
    
    it('callback is called', () => {
        const item='Lorem ipsum dolor sit amet consectetur'
        const { getByTestId } = render(
            <ListItem
                id='1'
                checkable={true}
                onCheck={mockOnCheck}
                item={item}
            />
        );
        const input = getByTestId("test-1");
        fireEvent.click(input);
        expect(mockOnCheck).toHaveBeenCalledTimes(1);
    });

    it('callback is not called when not checkable', () => {
        const mockCallback = jest.fn();
        render(<ListItem checkable={false} onCheck={mockCallback} id="test-1" item="test item" />);
        const checkbox = screen.queryByTestId('test-test-1');
        if (checkbox) {
            fireEvent.click(checkbox);
        }
        expect(mockCallback).not.toHaveBeenCalled();
    });

    it('matches saved snapshot', () => {
        const tree = render(
            <ListItem
                id='list-item-test'
                checkable={true}
                onCheck={mockOnCheck}
                item='Lorem ipsum dolor sit amet consectetur'
            />
        );
        expect(tree).toMatchSnapshot();
    });
});