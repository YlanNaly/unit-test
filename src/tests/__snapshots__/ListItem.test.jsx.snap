// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem matches saved snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="list-item"
        data-testid="test-list-item-test-container"
      >
        <input
          data-testid="test-list-item-test"
          id="list-item-test"
          type="checkbox"
        />
        <label
          data-testid="label-item"
          for="list-item-test"
        >
          Lorem ipsum dolor sit amet consectetur
        </label>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="list-item"
      data-testid="test-list-item-test-container"
    >
      <input
        data-testid="test-list-item-test"
        id="list-item-test"
        type="checkbox"
      />
      <label
        data-testid="label-item"
        for="list-item-test"
      >
        Lorem ipsum dolor sit amet consectetur
      </label>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
